File: src/lib/progressiveOverload.ts
import type { WorkoutLog } from '../store/useAppStore';

export interface ProgressiveSuggestion {
  suggestedWeight: number;
  currentWeight: number;
  sessionsAnalyzed: number;
  increment: number;
}

/**
 * Analyzes the last `minSessions` distinct training sessions for an exercise.
 * Returns a weight suggestion when the user consistently hit the same peak weight.
 */
export function getProgressiveSuggestion(
  exerciseId: string,
  logs: WorkoutLog[],
  minSessions = 3,
): ProgressiveSuggestion | null {
  const exerciseLogs = logs
    .filter((l) => l.exerciseId === exerciseId)
    .sort((a, b) => b.date.localeCompare(a.date));

  if (exerciseLogs.length === 0) return null;

  // Group logs by date (YYYY-MM-DD)
  const sessionMap = new Map<string, WorkoutLog[]>();
  for (const log of exerciseLogs) {
    const date = log.date.slice(0, 10);
    if (!sessionMap.has(date)) sessionMap.set(date, []);
    sessionMap.get(date)!.push(log);
  }

  // Most-recent sessions first
  const sessions = Array.from(sessionMap.entries())
    .sort(([a], [b]) => b.localeCompare(a))
    .slice(0, minSessions);

  if (sessions.length < minSessions) return null;

  // Max weight per session
  const weights = sessions.map(([, s]) => Math.max(...s.map((l) => l.weight)));

  // All sessions must share the same peak weight
  if (!weights.every((w) => w === weights[0])) return null;

  const currentWeight = weights[0];
  if (currentWeight <= 0) return null;

  const increment = 2.5;
  // Round to nearest 0.5 kg (standard gym plate increment)
  const suggestedWeight = Math.round((currentWeight + increment) * 2) / 2;

  return { suggestedWeight, currentWeight, sessionsAnalyzed: sessions.length, increment };
}

export interface ExerciseSession {
  date: string; // YYYY-MM-DD
  maxWeight: number;
  totalVolume: number; // sum of weight × reps
  sets: number;
  maxReps: number;
}

/**
 * Returns per-session aggregates for an exercise, sorted oldest→newest (for chart).
 */
export function getExerciseHistory(
  exerciseId: string,
  logs: WorkoutLog[],
  limit = 15,
): ExerciseSession[] {
  const exerciseLogs = logs.filter((l) => l.exerciseId === exerciseId);
  const sessionMap = new Map<string, WorkoutLog[]>();

  for (const log of exerciseLogs) {
    const date = log.date.slice(0, 10);
    if (!sessionMap.has(date)) sessionMap.set(date, []);
    sessionMap.get(date)!.push(log);
  }

  return Array.from(sessionMap.entries())
    .sort(([a], [b]) => a.localeCompare(b))
    .slice(-limit)
    .map(([date, s]) => ({
      date,
      maxWeight: Math.max(...s.map((l) => l.weight)),
      totalVolume: s.reduce((sum, l) => sum + l.weight * l.reps, 0),
      sets: s.length,
      maxReps: Math.max(...s.map((l) => l.reps)),
    }));
}

File: voltbody-android/app/src/main/java/com/voltbody/app/domain/usecase/ProgressiveOverload.kt
package com.voltbody.app.domain.usecase

import com.voltbody.app.domain.model.WorkoutLog

// ── Progressive Overload (ported from progressiveOverload.ts) ─────────────────

data class ProgressiveSuggestion(
    val suggestedWeight: Float,
    val currentWeight: Float,
    val sessionsAnalyzed: Int,
    val increment: Float
)

fun getProgressiveSuggestion(
    exerciseId: String,
    logs: List<WorkoutLog>,
    minSessions: Int = 3
): ProgressiveSuggestion? {
    val exerciseLogs = logs.filter { it.exerciseId == exerciseId }.sortedByDescending { it.date }
    if (exerciseLogs.isEmpty()) return null

    val sessionMap = mutableMapOf<String, MutableList<WorkoutLog>>()
    exerciseLogs.forEach { log ->
        val date = log.date.take(10)
        sessionMap.getOrPut(date) { mutableListOf() }.add(log)
    }

    val sessions = sessionMap.entries.sortedByDescending { it.key }.take(minSessions)
    if (sessions.size < minSessions) return null

    val weights = sessions.map { (_, s) -> s.maxOf { it.weight } }
    if (weights.any { it != weights[0] }) return null

    val currentWeight = weights[0]
    if (currentWeight <= 0f) return null

    val increment = 2.5f
    val suggestedWeight = (Math.round((currentWeight + increment) * 2) / 2).toFloat()
    return ProgressiveSuggestion(suggestedWeight, currentWeight, sessions.size, increment)
}

data class ExerciseSession(
    val date: String,
    val maxWeight: Float,
    val totalVolume: Float,
    val sets: Int,
    val maxReps: Int
)

fun getExerciseHistory(
    exerciseId: String,
    logs: List<WorkoutLog>,
    limit: Int = 15
): List<ExerciseSession> {
    val exerciseLogs = logs.filter { it.exerciseId == exerciseId }
    val sessionMap = mutableMapOf<String, MutableList<WorkoutLog>>()
    exerciseLogs.forEach { log ->
        val date = log.date.take(10)
        sessionMap.getOrPut(date) { mutableListOf() }.add(log)
    }
    return sessionMap.entries
        .sortedBy { it.key }
        .takeLast(limit)
        .map { (date, s) ->
            ExerciseSession(
                date = date,
                maxWeight = s.maxOf { it.weight },
                totalVolume = s.sumOf { it.weight * it.reps }.toFloat(),
                sets = s.size,
                maxReps = s.maxOf { it.reps }
            )
        }
}
